### PR TITLE
Organizers should be able to connect to the PayPal Checkout gateway

### LINKED ISSUE #4892: Organizers should be able to connect to the PayPal Checkout gateway
## User Story

As an organizer, I want to connect my PayPal account to the PayPal Checkout gateway so that I can start receiving donations via that gateway.

## Details
This piece of the PayPal Commerce integration should allow users to connect their PayPal account to the PayPal Checkout gateway.

## Visuals
PayPal Checkout _ Not Connected
PayPal Checkout _ Connected

## Acceptance Criteria

- [x] PayPal Checkout connect settings are available
- [x] User is able to connect a new account to PayPal Checkout
- [x] User is able to connect an existing account to PayPal Checkout
- [x] User is able to disconnect their account

### PR DESCRIPTION
**Pull request by ravinderk**
Resolves #4892

## Description

This PR adds logic to connect the Paypal account to process donations with PayPal commerce.

## Affects

I added the `boot` method to `interface PaymentGateway` which required because I need that function setup hooks.

## Pre-review Checklist

-   [x] Acceptance criteria satisfied and marked in related issue
-   [ ] Tests included
-   [ ] Keyboard accessible
-   [ ] Screen reader accessible
-   [x] Relevant `@since` tags included in DocBlocks
-   [ ] Changelog updated
-   [ ] Labels applied if docs are needed
-   [x] [Self Review](https://give.gitbook.io/development-manual/devops/github/code-reviews#self-review) of code and UX completed

## Testing Instructions

- [ ] can you connect the PayPal account?
- [ ] Are Paypal account actions working fine for a connect PayPal account?

### CODE REVIEW

**Comment by JasonTheAdams** [comment]
@ravinderk This seems broken for me. I ran `composer install`, `npm run dev`, and refreshed my permalinks.
Error

**Comment by ravinderk** [comment]
As I discussed with @JasonTheAdams this pr logic works fine if admin set GiveWP to test mode. @JasonTheAdams will look into issues we are facing when test mode disabled.

**Comment by ravinderk** [comment]
This pr is incomplete, so moving this back to `Programming`. I discovered that have to need to store `access_token` for the future purpose which we got during the onboarding process.

Slack discussion: https://givewp.slack.com/archives/C0FAGC83C/p1595267338013100

**Comment by JasonTheAdams** [comment]
## Call Summary
**Attendees:** Jason Adams & Ravinder
**Subject:** Token Refreshing

Ravinder is wrapping up the onboarding PR, part of which is handling token refreshing. We discussed some strategies for this and their various pros/cons, settling on his implementation using the cron job to mark expiration and hopefully anticipate it.

I also used the wrong hook (an uncached hook) which was causing the gateway to be registered multiple times, causing chaos. Hahah! Turns out we have more than one hook for registering gateways... which is confusing. But we're using the right hook now and I've learned something.

**Comment by JasonTheAdams** [comment]
@ravinderk Is this ready to be reviewed again?

**Comment by ravinderk** [comment]
> @ravinderk Is this ready to be reviewed again?

Yes

**Comment by ravinderk** [comment]
After discussion with PayPalTeam,  @JasonTheAdams and I find out that we need to develop two API endpoints on the partner server to get client rest API endpoint and onboarding status.

@JasonTheAdams will developer this endpoint and till then he is holding a review. I will integrate endpoint to the onboarding process after he will complete partner server endpoints development.

**Comment by ravinderk** [comment]
@JasonTheAdams this pr is ready to review and I uploading [Gateway OAuth PR](https://github.com/impress-org/gateway-oauth/pull/9) to testing server to test this functionality.

File: src/PaymentGateways/PayPalCommerce/MerchantDetail.php [hunk 1]
```diff
@@ -0,0 +1,215 @@
...
+	 * @param array $merchantDetails
+	 *
+	 * @since 2.8.0
+	 *
+	 * @return MerchantDetail
+	 */
+	public function fromArray( $merchantDetails ) {
+		$obj = new static();
+
+		$obj->setupProperties( $merchantDetails );
```
Comment (JasonTheAdams) [inline src/PaymentGateways/PayPalCommerce/MerchantDetail.php:148]: Make sure to validate here, as well.

File: src/PaymentGateways/PayPalCommerce/onBoardingRedirectHandler.php [hunk 2]
```diff
@@ -0,0 +1,252 @@
...
+		$required      = [ 'payments_receivable', 'primary_email_confirmed' ];
+		$onBoardedData = array_filter( $onBoardedData ); // Remove empty values.
+
+		if ( array_diff( $required, array_keys( $onBoardedData ) ) ) {
+			$this->redirectWhenOnBoardingFail();
+		} elseif ( ! $onBoardedData['payments_receivable'] ) {
+			$this->redirectWhenOnBoardingFail( esc_html__( 'We are unable to connect your seller account because your seller account can not receive payments. Please contact PayPal Support Team', 'give' ) );
+		} elseif ( ! $onBoardedData['primary_email_confirmed'] ) {
+			$this->redirectWhenOnBoardingFail( esc_html__( 'We are unable to connect your seller account because your seller account primary email has not been confirmed. Please contact PayPal Support Team', 'give' ) );
+		}
```
Comment (JasonTheAdams) [inline src/PaymentGateways/PayPalCommerce/onBoardingRedirectHandler.php:228]: I don't _think_ we're saying that they didn't successfully connect, but that they need to do things on their account for it to be workably ready. I'm thinking we just append a list of needs something like:
```
Your are successfully connected, but you need to do a few things within your PayPal account before you're ready to receive donations:
- Confirm your primary email address
- Set up an account to receive payment from PayPal
```
Just have the missing items append to the list. That make sense?

File: src/PaymentGateways/PayPalCommerce/MerchantDetail.php (hunk 1 above)
Comment (ravinderk) [reply src/PaymentGateways/PayPalCommerce/MerchantDetail.php:149]: fixed: https://github.com/impress-org/givewp/compare/4eb148d59e6a...0e0e27168d4b

File: src/PaymentGateways/PayPalCommerce/onBoardingRedirectHandler.php (hunk 2 above)
Comment (ravinderk) [reply src/PaymentGateways/PayPalCommerce/onBoardingRedirectHandler.php]: fixed: https://github.com/impress-org/givewp/compare/0e0e27168d4b...94c267311362

image

**Comment by ravinderk** [comment]
@jason I updated pr as per your suggestion. PR ready to review.

**Pull request merged**